### DISCUSSION TITLE
RESTful에 대해서 설명해주세요.

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#172] [Network]
.

**Comment by gyungchan-jo** [comment]
RESTful은 REST 원칙을 준수하여 웹 서비스를 설계하는 스타일입니다.

REST의 여섯 가지 기본 원칙:

1. 일관된 인터페이스 (Uniform Interface) 
: 일관된 인터페이스를 통해 시스템 아키텍처를 단순화하고 상호작용의 가시성을 높입니다.

2. 클라이언트-서버 (Client-Server) 
: 클라이언트와 서버는 명확하게 역할이 분리됩니다. 클라이언트는 자원을 요청하고, 서버는 요청을 처리하여 응답을 반환합니다.

3. 무상태성 (Stateless) 
: 클라이언트와 서버 간의 모든 요청은 독립적이며, 서버는 각 요청을 별도로 처리합니다. 서버는 클라이언트의 상태를 저장하지 않고, 필요한 모든 정보는 요청에 포함되어야 합니다.

4. 캐시 가능 (Cacheable) 
: 응답 데이터는 캐시될 수 있으며, 이를 통해 성능을 최적화할 수 있습니다.

5. 계층화 시스템 (Layered System) 
: RESTful 시스템은 여러 계층으로 구성될 수 있으며, 각 계층은 독립적으로 기능할 수 있습니다. 예를 들어, 로드 밸런서, 캐시 서버, 애플리케이션 서버 등이 계층화 시스템의 예입니다.

6. 코드 온 디맨드 (선택적) (Code on Demand) 
: 서버가 클라이언트에서 실행시킬 수 있는 로직을 전송하여 클라이언트의 기능을 확장시킬 수 있다. 이를 통해 클라이언트가 사전에 구현해야하는 기능의 수를 줄여 간소화시킬 수 있습니다.

[참고 문서](https://restfulapi.net/)

**Comment by nayonsoso** [comment]
### REST / RESTful / REST API
- REST : 소프트웨어 아키텍처 스타일
- RESTful 웹 서비스 : Rest 아키텍처를 구현하는 웹서비스
- REST API (=Restful API) : Rest 아키텍처 스타일을 따르는 API
- 따라서 Restful 이 무엇인지 알기 위해서는 Rest 아키텍처에 대해 알아야 한다.
- 말장난같지만 출처는 aws (https://aws.amazon.com/ko/what-is/restful-api/)

### REST 아키텍처의 원칙
1. 일관된 인터페이스
- 요청과 응답은 표준 형식이어야 한다. 
- 요청은 리소스를 식별해야 한다. 이를 위해서 각 리소스는 고유한 식별자(URI)를 가져야 한다.
- 요청은 리소스를 어떻게 사용할 것인지가 표함된 메세지(e.g HTTP 메서드)를 포함해야 한다.
- 요청이 제대로 이루어질 수 있도록 서버는 다른 리소스들에 관한 충분한 정보를 제공해야 한다. 
  - e.g. 사용자가 특정 테마를 선택해 예약을 할 수 있도록 테마의 id 를 제공한다.
- ⚠️ Self-descriptive messages (자기 서술적 메세지) : 메세지는 스스로를 설명해야 한다.
- ⚠️ Hypermedia as the engine of application state(HATEOAS) : 어플리케이션의 상태는 Hyperlink 를 통해 전이되어야 한다.
  - cf. ⚠️로 표시한 조건들 때문에 우리가 소위 ’REST api’ 라고 부르는 대부분의 api들이 엄밀히 따지자면 REST 하지 않다고 할 수 있다. 
  - 자세한 건 여기 : https://www.youtube.com/watch?v=RP_f5dMoHFc&t=5s / https://velog.io/@kjh03160/그런-REST-API로-괜찮은가.

2. 무상태
- 서버는 클라이언트의 정보를 저장하지 않는다.
- 모든 요청은 서버가 요청을 처리하는데에 필요한 모든 정보를 포함해야 한다.
- 따라서 각 요청은 다른 요청들과 독립적으로 처리될 수 있게 한다.

3. 계층화
- 서버와 클라이언트 사이에는 다른 중개자가 있을 수 있다.
- 또한 서버가 다른 서버로 요청을 보내도록 설계할 수 있다.
- 이런 계층은 클라이언트에 보이지 않는 상태로 유지된다.

4. 캐시 가능성
- RESTful 웹 서비스는 서버 응답 시간을 개선하기 위해 캐싱을 지원한다.
- 따라서 응답은 캐시가 가능하거나 불가능하다는 것을 포함해야 한다.
